project_milestone/tests/test_project_milestone.py

```diff
@@ -0,0 +1,59 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo.tests import common
+
+
+class TestProjectMilestone(common.TransactionCase):
+
+    def setUp(self):
+        super().setUp()
+        self.test_project = self.env['project.project'].create({
+            'name': 'TestProject'})
+        self.test_project_milestone_1 = self.env['project.milestone'].create({
+            'name': 'TestMilestone_1',
+            'project_id': self.test_project.id})
+        self.test_project_milestone_2 = self.env['project.milestone'].create({
+            'name': 'TestMilestone_2',
+            'project_id': self.test_project.id})
+        self.test_open_stage = self.env['project.task.type'].create({
+            'name': 'TestOpenStage'})
+        self.test_close_stage = self.env['project.task.type'].create({
+            'name': 'TestCloseStage',
+            'closed': True})
+        self.test_task = self.env['project.task'].create({
+            'name': 'TestTask',
+            'project_id': self.test_project.id,
+            'milestone_id': self.test_project_milestone_1.id,
+            'stage_id': self.test_open_stage.id})
+        self.env['project.task'].create({
+            'name': 'TestTask',
+            'project_id': self.test_project.id,
+            'milestone_id': self.test_project_milestone_1.id,
+            'stage_id': self.test_close_stage.id})
+
+    def test_milestone_sequences(self):
+        milestone1 = self.test_project_milestone_1
+        milestone2 = self.test_project_milestone_2
+
+        self.assertGreater(milestone2.sequence, milestone1.sequence)
+
+    def test_milestone_progress(self):
+        milestone1 = self.test_project_milestone_1
+
+        self.assertEqual(milestone1.progress, 50)
+
+    def test_read_group_milestone_ids(self):
+        # Add the default_project_id in the context and prep the arguments
+        milestone_ids = self.env['project.milestone'].search([(
+            'project_id', '=', self.test_project.id)])
+        domain = [("project_id", "=", self.test_project.id)]
+        order = 'sequence'
+
+        project_01 = self.test_project
+        # Call the method to test
+        res = project_01.task_ids._read_group_milestone_ids(milestone_ids,

```

Comment:
Better call directly `search_read()` as I demonstrated above. This is the method that is called from the web client, and you want to make sure it behaves properly. It will call your `_read_group_milestone_ids()` behind the scenes if everything is properly done (it is).

Also, you don't call that method usually from a recordset, but from the model itself, since it's used to actually find records.

So, this should look like:

```python
Task = self.env["project.task"]
grouped_tasks = Task.search_read(domain, ["name"], ["milestone_id"])
self.assertEqual(len(grouped_tasks), 1)
```

You noticed I expect just 1 group. That's because all tasks that are going to be found are in the milestone 1.

Now, when you set `default_project_id` in the context, it should return 2 results, although the 2nd will be empty. You just alter the context, search again and assert the results length:

```python
Task = Task.with_context(default_project_id=self.test_project.id)
grouped_tasks = Task.search_read(domain, ["name"], ["milestone_id"])
self.assertEqual(len(grouped_tasks), 2)
```

It's most likely the code I recommended to you before, as you can see.

Reply:
I guess I didn't realize it would still call the function behind the scenes. Thank you for explaining this further. When I try to do this, it fails with the following errors:

2019-05-07 14:17:51,298 6571 ERROR openerp_test odoo.addons.project_milestone.tests.test_project_milestone: `     "Invalid value %r in domain term %r" % (right, leaf)
2019-05-07 14:17:51,298 6571 ERROR openerp_test odoo.addons.project_milestone.tests.test_project_milestone: ` AssertionError: Invalid value project.project(82,) in domain term ('project_id', '=', project.project(82,))

I thought maybe the domain needed to be [('project_id', '=', self.test_project.id)] instead of self.test_project but then it errors with something like expecting int and not list (not exact error, I should have copied it).

My googling is failing to help me understand the error, I'll continue to look but do you have any ideas?

Reply:
> I thought maybe the domain needed to be [('project_id', '=', self.test_project.id)] instead of self.test_project

Indeed!

> but then it errors with something like expecting int and not list (not exact error, I should have copied it).

Yes, copy it here please.